### DISCUSSION TITLE
Implement post tasks on project targets

### DISCUSSION THREAD

**Discussion by rpanadero** [nrwl/nx#27048] [Feature Requests]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
I would like to see Nx support post- tasks on project targets. Other tools like npm do implement this feature.

## Motivation
It is very usual to need to run a task after running a target, for example after running `build` target. Maybe not everyone needs this, but it would be really helpful to have it.

For example, I'm working on an Angular project right now where I need to instrument all generated js files after building for production by running a command. Now, to do it, I would need to add an additional target like "build-full", and rename all references of "build" target to the new "build-full"... Or another approach would be to implement a custom builder for my angular app and give it the ability to run a target after building. Both of them are really tedious, from my point of view. Also, please, consider that other tools like NPM do implement this feature (https://docs.npmjs.com/cli/v10/using-npm/scripts#pre--post-scripts). "DependsOn" works well for pre tasks, but we need something similar for post tasks.

I have seen that this feature request was also requested in the past and discarded, but I would appreciate you check it out again. Le me reference a past requests which was closed by inactivity:

https://github.com/nrwl/nx/issues/9054

## Suggested Implementation

```
"build": {
  "executor": "@angular-devkit/build-angular:browser",
  "postTargets": ["postbuild", ...],
  "options": {
     ...
   }
},
"postbuild": {
  "executor": "@nrwl/workspace:run-commands",
  "options": {
      "commands": ["echo 'TEST'"],
   }
}
```

`postTargets` syntax would be equal to `dependsOn`.

## Alternate Implementations

```
"build": {
  "executor": "@angular-devkit/build-angular:browser",
  "options": {
     postTargets: ["myapp:postbuild", ...]
   }
},
"postbuild": {
  "executor": "@nrwl/workspace:run-commands",
  "options": {
      "commands": ["echo 'TEST'"],
   }
}
```

This approach would allow us to have different post targets by configuration. "PostTargets" option would be like a reserved option for an executor.

**Comment by AhsanAyaz** [comment]
Would love to see this being added 👍🏽

**Comment by jahusa02** [comment]
I have actually the same usecase. Need to upload sourcemaps to sentry and the esbuild plugin doesnt work with NX/Angular.

**Comment by rpanadero** [comment]
This feature request seems to have quite support from the community, could you have a look at it?

**Comment by joshuaben-seth** [comment]
This would be quite useful.

**Comment by Solat-Ali** [comment]
Any updates on this request?

**Comment by eggp** [comment]
It would be great to have this feature for Transloco as well, because in the pre-build phase, we need to call the transloco-scoped-libs, and after the build, we need to optimize with the transloco-optimize command. This would really help so that we don't have to write an npm post hook for every command.

**Comment by FranciscoKloganB** [comment]
+1

**Comment by FrozenPandaz** [comment]
Hey all, I wanted to discuss this more before we go and implement it.

I think the current configuration allows you to easily is to specify post targets: 
```
"poststuff": {
  "dependsOn": ["stuff"],
  "command": "echo hi"
},
```

Then to specify that you want both `poststuff` and `stuff` to run you can specify both as targets: `nx run-many -t stuff,poststuff`

What do you all think of this solution? Does it solve your problems? Is it intuitive?

**Reply by FranciscoKloganB** [reply]
I think `poststuff` should run after `stuff` without having to explicitly having to `run-many` (autorun behaviour could be opt-in).

I'm a bit new to Nx, but my gripe with having to use `run-many` is that it sometimes I may want some `poststuff` to run after `stuff` on project `foo`, before `stuff` runs on project `bar`, if `bar` depends on `foo`. Another issue with `run-many` is that I want teams to know as few commands as possible.

**Reply by tukusejssirs** [reply]
@FrozenPandaz, I used that ATM, however, I consider it a workaround. I prefer to have a way to specify explicitly what should be run after the current job, similarly how it can be done in NPM/PNPM scripts (although implicitly).

There are some after tasks (`poststuff`) that can be by themselves, however, the `stuff` task always requires `poststuff` to be run after `stuff`. Moreover, it is mentally more preferable to me to define which tasks should be executed after the currwnt task than using `dependsOn`.

Consider adding `runAfter` (name it whatever you want).

**Comment by rpanadero** [comment]
Any updates on this request?

**Comment by csaszika** [comment]
@FrozenPandaz any updates on this request? It would be really useful.

**Comment by kevinswarner** [comment]
So, somewhat similar to this is an approach I am just starting to look into. I have several targets (tasks) that all have a need to have one or more local services started (e.g. think docker), and also have one or more resources created within those services (e.g. think a database, or an queue.) At first, I thought about just creating a separate "services" project, and a "resources" project, and then have my other tasks "dependOn" them. But, I want to be able to have a little more control over those tasks on which another task depends. For example, one task may require the services "postgres" and "redis", while another may only require "postgres". So, being able to specify the services would be nice. Likewise, a specific task may need a database initialized with a certain schema, and a S3 bucket created. But another task may only need a bucket. I am simplifying this a little bit just for discussion.

So, my initial thought was to create a custom executor, similar to the "nx:run-commands" executor, but that accepts additional options. Maybe a services array, and a resources array. Again, over simplified. But, the idea is that the custom executor would basically wrap the nx:run-commands executor so I could use it like this...

```
{
  "targets": {
    "test": {
      "executor": "@my-plugin/:run-commands",
      "options": {
        "services": ["postgres", "redis"],
        "resources": ["foo", "bar"],
        "commands": ["my-test-script-or-command"]
      }
    }
  }
}
```

Anyway, I have not gotten far, and this may not be the most idiomatic approach. But, it could be used for pre or post target actions.

**Reply by eggp** [reply]
```
  "targets": {
    "pre-build": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm run transloco:extract-scoped-libs"
      }
    },
    "build": {
      "dependsOn": [
        {
          "target": "pre-build",
          "projects": "self"
        }
      ],.....
```

This works for me in an Angular environment, maybe it can help you too. Although I would also expect a cleaner solution, this is what it is. :) :(

**Reply by kevinswarner** [reply]
Thanks. I also went down the path of using dependsOn, but I need to be able to pass arguments in the dependsOn, which is not possible, as far as I have been able to tell.

**Comment by kevinswarner** [comment]
I have been working down the path I described above, but I am having trouble "wrapping" the nx:run-commands executor. I cannot figure out a clean way to invoke that executor from my executor so that I get all of the benefits of Nx on the commands passed to my executor. I tried the runExecutor function, but I cannot figure out how to specify the nx:run-commands executor.

**Comment by marc-wilson** [comment]
It would be great to get this added. Most notable issue for me is that i need to copy some files and potentially do some things _after_ the dist is produced. But, I can't because I use docker.

For example:
```
nx run my-app:build
cp somefile.txt ./apps/my-app/dist
nx run my-app:docker:build
```

In this scenario, whatever you do before `docker:build` will get erased. `docker:build` (i assume) has `dependsOn: ["build"]` so whatever I do before `docker:build` is irrelevant. If there were a post-build action, this wouldn't be a problem. Maybe it's a bug that build is executed twice in this scenario, but I don't think so.

I am aware I can just create my own target that does both, but i'm trying to leverage the new inferred tasks.
